source/common/http/filter_manager.cc

```diff
@@ -862,6 +862,12 @@ void FilterManager::sendLocalReplyViaFilterChain(
   // a no-op.
   createFilterChain();
 
+  // A bit awkward, but save the local reply data for later. This will come in handy.
+  // When this pointer is non-null, it means a local reply has been initiated, which
+  // is important later on.
+  local_reply_data_ = std::make_unique<Utility::LocalReplyData>(
```

Comment:
I think the best explanation is that we may need some of this context if we end up doing the rewrite in encodeData (because an upstream response exists and needs to be dropped / replaced by the rewritten body, so this cannot happen in encodeHeaders alone)

Reply:
After reviewing again, I agree this is pretty dirty but it currently serves an important purpose. Mostly, to have the local reply code path communicate to the canonical rewrite path that the local reply is grpc, is a head request, the grpc status, local reply body etc. I think this would seem cleaner, at least on the surface, if there were one member `local_reply_context_` that held both local reply data (if any) and a pointer to a local reply object that should be used for the request. This could also give us an opportunity to read per-route config and override that local reply object to the per-route variant. The alternative would be to create the FM with the per-route local reply rewriter, but I'm not sure yet which makes the most sense from a lifecycle perspective.